Windows: build broken in `utils::fs::set_executable`

```
   Compiling jokeyrhyme-dotfiles v0.31.0
error[E0277]: can't compare `P` with `P`
   --> C:\Users\ronwa_op69a\.cargo\registry\src\github.com-1ecc6299db9ec823\jokeyrhyme-dotfiles-0.31.0\src\utils\github.rs:116:5
    |
116 |     utils::fs::set_executable(&bin_path)?;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^ no implementation for `P == P`
    |
    = help: the trait `std::cmp::PartialEq` is not implemented for `P`
    = help: consider adding a `where P: std::cmp::PartialEq` bound
    = note: required because of the requirements on the impl of `std::cmp::PartialEq` for `&P`
note: required by `utils::fs::set_executable`
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -28,7 +28,7 @@ install:
 script:
   - cargo build --verbose
   - cargo test
-  - cargo clippy --all-features --all-targets -- -D clippy::all -D clippy::pedantic
+  - cargo clippy --all-features --all-targets -- -D clippy::all
 
 deploy:
   on:

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,6 +2,13 @@ language: rust
 rust:
   - stable
 
+matrix:
+  include:
+    - os: linux
+    - os: osx
+      osx_image: xcode10.2
+    - os: windows
+
 # cache trick from https://github.com/japaric/trust
 before_cache:
   - chmod -R a+r $HOME/.cargo

--- a/.travis.yml
+++ b/.travis.yml
@@ -16,6 +16,7 @@ cache: cargo
 
 env:
   global:
+    - RUST_BACKTRACE=1
     - secure: GxL8AeWbzT2fTwWZ/RhFl3HMCGeHc5LHK+SYSdf3S0WfxIyg+1LsekNx+QNqqA1iJjLrmWoyDfGoRkS7/46hkRKZ+kci61xDIU4JcpW8hwo5ANjJzGq0x+LCekkGjwQeaG92D2rRFOX+IOEkSZKWCWHj/xkLCydLGBHcICFenC+jWRjifbwVj2b38tWxhKFIMaKuATtHRIU1FwzqVW8kE3ROiuSpEK6Guo7QNNqh0Fj+z/cYzBmiE8i7nFC19yAUyfs2Zjv2hUntQlORQhO7cIE2dUXSr/5B6zvy/1e6uBAPDO+hWn6mbSBHRIAs6GzPQV4kb6taNpih30g7g51OSW6TQxIcMZzmhsJDh48cTTqN9TKaF8hNsVxgOIk7zs+MurustCWXFZwpC+us2fTd9mYNloW/u+wrRAlXTzKQtbQn7CDnjX+C9bLtWrr6RnP2V8cdfLEI0+V9LN4uCP4xEpyQkyXX6kFZXcAoe4LFrezUkNyDKvarR59qRHCIfWCeOde+Icoh/fAHgtUawqDrTHWTy4GcGaaFzVwSBA36+IjH5lMoR2lCdwxTjk/wVFBGBKjwzqfxZSSwwx48sNQSt8H7QSvKMdqb98X2E4ZTrQnplT7tLMsdy8iF9VCbg4TqcDXeLcn5DE30yZIpTzrRPGlwHlBiKLdnUvJHrwe6wKY=
     - secure: VeIkqEoSNHEQaeXciAt+2WICvxc45xPscDwiZCxmuLOyYjUSxp1kazyoY9GF7NFqgkdmB9+0a5p9CJF7qVGSoaCMv0bUv9QIgZLOCay1j278v4nsOHj2dtdqIvoKLL6HEfVtS4LYoI5CdG6hNuck50eUjGC+Y/vkkhBfYVRueebh/kwDgjGKBdbnPPb2yFsKkf3NTFoGjjDXUtjlRNroD80bz1WVaOYCYhJZGzB2pf4eBTEPjFqvpv2pBjC/PwRA18rAX1PdpuQp1gUIL7Ig3IvRjP2111YCdHc2uq6w4DQKpOLJ24lEEpcP4r/DcpLg8IQpwHokvesN2cYLs7j71HLpRuOdVyHudDxnxHq36C8NyZ65lE/2ByXqclYqxtzjFDjHlXAsHlb7Yy331Zrp+v010hBIm5gC0O1Bpvn9hLuAYPqFxME7arYicDw0WLPNH5RAXlOsAxtG1H3T6CGRU4ClpNuJTN70XZ7NP9Hv4f6lsgZxSVprUikBBt/+Sb17XMzKAPtolY4NoQGKv8w8elh4q2EIRwR/udQWPq0WFlUl9YhnRic8jofHOS16wkxmdscUa0PWD5QhzTgbLM1tnl4ynaXfOdMlkz6ECIxRmOrcQFRxpvJhc+BBiWvvBh7QmQcehYA5waU+HipFIvme/uI2d0MbgkWsHFJ8+gkKm6U=
 
@@ -26,8 +27,8 @@ install:
 
 script:
   - cargo build --verbose
-  - make test
-  - make lint
+  - cargo test
+  - cargo clippy --all-features --all-targets -- -D clippy::all -D clippy::pedantic
 
 deploy:
   on:

--- a/.travis.yml
+++ b/.travis.yml
@@ -17,26 +17,15 @@ install:
 - rustup update
 - rustup component add clippy
 
-stages:
-  - name: test
-  - name: deploy
-    if: tag IS present AND fork = false
+script:
+  - cargo build --verbose
+  - make test
+  - make lint
 
-jobs:
-  include:
-    - name: build
-      stage: test
-      script: cargo build --verbose
-    - name: lint
-      script: make lint
-    - name: test
-      script: make test
-    - name: deploy to crates.io
-      stage: deploy
-      script: false
-      deploy:
-        on:
-          tags: true
-        provider: cargo
-        token: $CARGO_TOKEN
+deploy:
+  on:
+    repo: jokeyrhyme/dotfiles-rs
+    tags: true
+  provider: cargo
+  token: $CARGO_TOKEN
 

--- a/Makefile
+++ b/Makefile
@@ -6,9 +6,8 @@ fmt:
 .NOCACHE: lint
 lint:
 	set -eux
-	cargo fix --allow-dirty --allow-staged
 	touch src/main.rs && cargo check --all-features --all-targets
-	touch src/main.rs && cargo clippy --all-features --all-targets -- -D clippy::all
+	touch src/main.rs && cargo clippy --all-features --all-targets -- -D clippy::all -D clippy::pedantic
 
 .NOCACHE: test
 test:

--- a/Makefile
+++ b/Makefile
@@ -7,7 +7,7 @@ fmt:
 lint:
 	set -eux
 	touch src/main.rs && cargo check --all-features --all-targets
-	touch src/main.rs && cargo clippy --all-features --all-targets -- -D clippy::all -D clippy::pedantic
+	touch src/main.rs && cargo clippy --all-features --all-targets -- -D clippy::all
 
 .NOCACHE: test
 test:

--- a/src/lib/env.rs
+++ b/src/lib/env.rs
@@ -134,7 +134,11 @@ mod tests {
             path: vec![PathBuf::from("/usr/bin"), PathBuf::from("/bin")],
         };
         let got = exports.to_shell(Shell::Bash);
-        let want = "export EDITOR=/usr/bin/vim\n\n\nexport PATH=/usr/bin:/bin";
+        let want = if OS == "windows" {
+            "export EDITOR=/usr/bin/vim\n\n\nexport PATH=/usr/bin;/bin"
+        } else {
+            "export EDITOR=/usr/bin/vim\n\n\nexport PATH=/usr/bin:/bin"
+        };
         assert_eq!(got, want);
     }
 
@@ -147,7 +151,11 @@ mod tests {
             path: vec![PathBuf::from("/usr/bin"), PathBuf::from("/bin")],
         };
         let got = exports.to_shell(Shell::Fish);
-        let want = "set --export EDITOR /usr/bin/vim\n\n\nset --export PATH /usr/bin:/bin";
+        let want = if OS == "windows" {
+            "set --export EDITOR /usr/bin/vim\n\n\nset --export PATH /usr/bin;/bin"
+        } else {
+            "set --export EDITOR /usr/bin/vim\n\n\nset --export PATH /usr/bin:/bin"
+        };
         assert_eq!(got, want);
     }
 
@@ -160,7 +168,11 @@ mod tests {
             path: vec![PathBuf::from("/usr/bin"), PathBuf::from("/bin")],
         };
         let got = exports.to_shell(Shell::Zsh);
-        let want = "export EDITOR=/usr/bin/vim\n\n\nexport PATH=/usr/bin:/bin";
+        let want = if OS == "windows" {
+            "export EDITOR=/usr/bin/vim\n\n\nexport PATH=/usr/bin;/bin"
+        } else {
+            "export EDITOR=/usr/bin/vim\n\n\nexport PATH=/usr/bin:/bin"
+        };
         assert_eq!(got, want);
     }
 }

--- a/src/lib/ghratask.rs
+++ b/src/lib/ghratask.rs
@@ -29,7 +29,7 @@ impl<'a> GHRATask<'a> {
         let release = match self.latest_release() {
             Ok(r) => r,
             Err(error) => {
-                return Err(task::Error::GitHubError(
+                return Err(task::Error::GitHub(
                     format!("unable to check latest release for {:?}", &self.repo),
                     error,
                 ));

--- a/src/lib/ghrtask.rs
+++ b/src/lib/ghrtask.rs
@@ -54,7 +54,7 @@ impl<'a> GHRTask<'a> {
         let release = match self.latest_release() {
             Ok(r) => r,
             Err(error) => {
-                return Err(task::Error::GitHubError(
+                return Err(task::Error::GitHub(
                     format!("unable to check latest release for {:?}", &self.repo),
                     error,
                 ));

--- a/src/lib/pip.rs
+++ b/src/lib/pip.rs
@@ -57,15 +57,15 @@ impl Favourites for PipFavourites {
 }
 
 pub fn found_versions() -> Vec<PipPackage> {
-    if !has_pip() {
+    if !has() {
         return Vec::new();
     };
-    let stdout = pip_output(&["list", "--format=json", "--user"]).unwrap_or_default();
+    let stdout = exe_output(&["list", "--format=json", "--user"]).unwrap_or_default();
     parse_pippackages(stdout)
 }
 
-pub fn has_pip() -> bool {
-    pip_output(VERSION_ARGS).is_ok()
+pub fn has() -> bool {
+    exe_output(VERSION_ARGS).is_ok()
 }
 
 pub fn pip<S>(args: &[S]) -> io::Result<()>
@@ -75,7 +75,7 @@ where
     command_spawn_wait(pip_exe()?, args).map(|_| ())
 }
 
-pub fn pip_output<S>(args: &[S]) -> io::Result<String>
+pub fn exe_output<S>(args: &[S]) -> io::Result<String>
 where
     S: AsRef<str>,
 {

--- a/src/lib/python.rs
+++ b/src/lib/python.rs
@@ -6,8 +6,8 @@ use which;
 const VERSION_ARGS: &[&str] = &["--version"];
 
 /*
-pub fn has_python() -> bool {
-    python_output(VERSION_ARGS).is_ok()
+pub fn has() -> bool {
+    exe_output(VERSION_ARGS).is_ok()
 }
 */
 
@@ -28,7 +28,7 @@ where
 }
 */
 
-pub fn python_output<O>(args: &[O]) -> subprocess::Result<String>
+pub fn exe_output<O>(args: &[O]) -> subprocess::Result<String>
 where
     O: AsRef<OsStr>,
 {

--- a/src/lib/task.rs
+++ b/src/lib/task.rs
@@ -7,37 +7,35 @@ use crate::utils::github;
 
 #[derive(Debug)]
 pub enum Error {
-    GitHubError(String, github::GitHubError),
-    IoError(String, io::Error),
-    NoTagsError,
-    PopenError(String, subprocess::PopenError),
+    GitHub(String, github::GitHubError),
+    Io(String, io::Error),
+    NoTags,
+    Popen(String, subprocess::PopenError),
 }
 impl std::error::Error for Error {}
 impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match &self {
-            Error::GitHubError(msg, cause) => {
-                write!(f, "{}", format!("{}: {:?}", msg, cause).red())
-            }
-            Error::IoError(msg, cause) => write!(f, "{}", format!("{}: {:?}", msg, cause).red()),
-            Error::NoTagsError => write!(f, "{}", "NoTagsError".red()),
-            Error::PopenError(msg, cause) => write!(f, "{}", format!("{}: {:?}", msg, cause).red()),
+            Error::GitHub(msg, cause) => write!(f, "{}", format!("{}: {:?}", msg, cause).red()),
+            Error::Io(msg, cause) => write!(f, "{}", format!("{}: {:?}", msg, cause).red()),
+            Error::NoTags => write!(f, "{}", "NoTagsError".red()),
+            Error::Popen(msg, cause) => write!(f, "{}", format!("{}: {:?}", msg, cause).red()),
         }
     }
 }
 impl From<github::GitHubError> for Error {
     fn from(cause: github::GitHubError) -> Error {
-        Error::GitHubError(String::new(), cause)
+        Error::GitHub(String::new(), cause)
     }
 }
 impl From<io::Error> for Error {
     fn from(cause: io::Error) -> Error {
-        Error::IoError(String::new(), cause)
+        Error::Io(String::new(), cause)
     }
 }
 impl From<subprocess::PopenError> for Error {
     fn from(cause: subprocess::PopenError) -> Error {
-        Error::PopenError(String::new(), cause)
+        Error::Popen(String::new(), cause)
     }
 }
 

--- a/src/tasks/dotfiles.rs
+++ b/src/tasks/dotfiles.rs
@@ -12,7 +12,7 @@ pub fn task() -> Task {
 fn sync() -> task::Result {
     let target = utils::env::home_dir().join(".dotfiles");
 
-    if utils::git::has_git() && utils::git::path_is_git_repository(&target) {
+    if utils::git::has() && utils::git::path_is_git_repository(&target) {
         utils::git::pull(&target);
     }
     Ok(Status::Done)

--- a/src/tasks/git.rs
+++ b/src/tasks/git.rs
@@ -86,7 +86,7 @@ where
 }
 
 fn sync() -> task::Result {
-    if !utils::git::has_git() {
+    if !utils::git::has() {
         return Ok(Status::Skipped);
     }
 

--- a/src/tasks/nodejs.rs
+++ b/src/tasks/nodejs.rs
@@ -81,7 +81,7 @@ fn sync() -> task::Result {
     let latest = utils::nodejs::latest_version();
     match install_nodejs(latest.clone()) {
         Ok(()) => Ok(Status::Changed(String::from("unknown"), latest)),
-        Err(error) => Err(task::Error::IoError(
+        Err(error) => Err(task::Error::Io(
             String::from("unable to install Node.js"),
             error,
         )),
@@ -101,7 +101,7 @@ fn update() -> task::Result {
     } else {
         match install_nodejs(latest.clone()) {
             Ok(()) => Ok(Status::Changed(current, latest)),
-            Err(error) => Err(task::Error::IoError(
+            Err(error) => Err(task::Error::Io(
                 String::from("unable to install Node.js"),
                 error,
             )),

--- a/src/tasks/pip.rs
+++ b/src/tasks/pip.rs
@@ -7,7 +7,7 @@ use crate::{
         env::Exports,
         favourites::Favourites,
         pip::{self, PipFavourites},
-        python::python_output,
+        python::exe_output,
         task::{self, Status, Task},
     },
     utils,
@@ -17,7 +17,7 @@ use crate::{
 // python -c "import site; print(site.USER_SITE)"
 
 pub fn env(mut exports: Exports) -> Exports {
-    let user_base = match python_output(&["-c", "import site; print(site.USER_BASE)"]) {
+    let user_base = match exe_output(&["-c", "import site; print(site.USER_BASE)"]) {
         Ok(ub) => ub,
         Err(_) => return exports,
     };
@@ -46,7 +46,7 @@ fn read_config() -> io::Result<PipFavourites> {
 }
 
 fn sync() -> task::Result {
-    if !pip::has_pip() {
+    if !pip::has() {
         return Ok(Status::Skipped);
     }
 
@@ -59,7 +59,7 @@ fn sync() -> task::Result {
 }
 
 fn update() -> task::Result {
-    if !pip::has_pip() {
+    if !pip::has() {
         return Ok(Status::Done);
     }
 

--- a/src/tasks/rust.rs
+++ b/src/tasks/rust.rs
@@ -78,7 +78,7 @@ fn sync() -> task::Result {
     let contents = match fs::read_to_string(&cfg_path) {
         Ok(s) => s,
         Err(error) => {
-            return Err(task::Error::IoError(String::from("ignoring config"), error));
+            return Err(task::Error::Io(String::from("ignoring config"), error));
         }
     };
 

--- a/src/tasks/ssh.rs
+++ b/src/tasks/ssh.rs
@@ -105,7 +105,7 @@ fn supported_ssh_macs() -> Vec<String> {
 }
 
 fn sync() -> task::Result {
-    if !utils::ssh::has_ssh() {
+    if !utils::ssh::has() {
         return Ok(Status::Skipped);
     }
 

--- a/src/tasks/vscode.rs
+++ b/src/tasks/vscode.rs
@@ -89,7 +89,7 @@ fn sync() -> task::Result {
     let contents = match fs::read_to_string(&cfg_path) {
         Ok(s) => s,
         Err(error) => {
-            return Err(task::Error::IoError(String::from("ignoring config"), error));
+            return Err(task::Error::Io(String::from("ignoring config"), error));
         }
     };
 

--- a/src/utils/fs.rs
+++ b/src/utils/fs.rs
@@ -1,13 +1,11 @@
-use std;
-#[cfg(unix)]
-use std::fs::File;
-#[cfg(unix)]
-use std::os::unix::fs::PermissionsExt;
 use std::{
+    self,
     fmt::Debug,
     io,
     path::{Path, PathBuf},
 };
+#[cfg(unix)]
+use std::{fs::File, os::unix::fs::PermissionsExt};
 
 use mktemp;
 
@@ -45,7 +43,6 @@ where
 }
 
 #[cfg(unix)]
-
 pub fn set_executable<P>(target: P) -> std::io::Result<()>
 where
     P: AsRef<Path>,
@@ -57,10 +54,9 @@ where
 }
 
 #[cfg(not(unix))]
-
 pub fn set_executable<P>(_target: P) -> std::io::Result<()>
 where
-    P: AsRef<Path> + PartialEq,
+    P: AsRef<Path>,
 {
     Ok(())
 }

--- a/src/utils/git.rs
+++ b/src/utils/git.rs
@@ -2,7 +2,7 @@ use std::{io, path::Path};
 
 use crate::utils;
 
-pub fn has_git() -> bool {
+pub fn has() -> bool {
     match utils::process::command_output("git", &["--version"]) {
         Ok(output) => output.status.success(),
         Err(_error) => false,
@@ -76,15 +76,15 @@ mod tests {
 
     #[test]
     #[cfg(unix)]
-    fn has_git_is_true_for_unix() {
-        let got = has_git();
+    fn has_is_true_for_unix() {
+        let got = has();
 
         assert!(got);
     }
 
     #[test]
     fn path_is_git_repository_here() {
-        if !has_git() {
+        if !has() {
             return;
         }
 
@@ -95,7 +95,7 @@ mod tests {
 
     #[test]
     fn shallow_clone_and_shallow_fetch() {
-        if !has_git() {
+        if !has() {
             return;
         }
 

--- a/src/utils/golang.rs
+++ b/src/utils/golang.rs
@@ -57,11 +57,11 @@ pub fn latest_version() -> Result<String, task::Error> {
         .collect();
 
     if release_tags.is_empty() {
-        return Err(task::Error::NoTagsError {});
+        return Err(task::Error::NoTags {});
     }
     match release_tags.last() {
         Some(latest) => Ok(latest.clone().id),
-        None => Err(task::Error::NoTagsError {}),
+        None => Err(task::Error::NoTags {}),
     }
 }
 

--- a/src/utils/ssh.rs
+++ b/src/utils/ssh.rs
@@ -661,6 +661,7 @@ mod tests {
             fs::read_to_string(&config_path)
                 .unwrap()
                 .replace("IdentityFile ~/.ssh/id_rsa", "IdentityFile ~\\.ssh\\id_rsa")
+                .replace("\r\n", "\n")
         } else {
             fs::read_to_string(&config_path).unwrap()
         };

--- a/src/utils/ssh.rs
+++ b/src/utils/ssh.rs
@@ -540,7 +540,7 @@ fn format_strings(key: &str, value: &Option<Vec<String>>) -> String {
     }
 }
 
-pub fn has_ssh() -> bool {
+pub fn has() -> bool {
     which::which("ssh").is_ok()
 }
 
@@ -577,10 +577,10 @@ where
             .split(',')
             .filter_map(|s| {
                 let trimmed = s.trim();
-                if !trimmed.is_empty() {
-                    Some(String::from(s))
-                } else {
+                if trimmed.is_empty() {
                     None
+                } else {
+                    Some(String::from(s))
                 }
             })
             .collect(),

